cmd/util/ledger/migrations/cadence_data_validation.go

```diff
@@ -0,0 +1,329 @@
+	m.rw = m.rwf.ReportWriter("cadence_data_validation")
+
+	return nil
+}
+
+func (m *postMigration) MigrateAccount(
+	ctx context.Context,
+	address common.Address,
+	payloads []*ledger.Payload,
+) ([]*ledger.Payload, error) {
+	newHash, err := m.v.hashAccountCadenceValues(address, payloads)
+	if err != nil {
+		m.log.Info().
+			Err(err).
+			Hex("address", address[:]).
+			Msg("failed to hash cadence values")
+		return payloads, nil
+	}
+
+	accountHash, ok := m.v.getAccountHash(address)
+
+	if !ok {
+		m.log.Error().
+			Hex("address", address[:]).
+			Msg("cadence values missing before migration")
+
+		m.rw.Write(
+			cadenceDataValidationReportEntry{
+
+				Address: address.Hex(),
+				Problem: "cadence values missing before migration",
+			},
+		)
+	}
+	if !bytes.Equal(accountHash, newHash) {
+		m.log.Error().
+			Hex("address", address[:]).
+			Msg("cadence values mismatch")
+
+		m.rw.Write(
+			cadenceDataValidationReportEntry{
+
+				Address: address.Hex(),
+				Problem: "cadence values mismatch",
+			},
+		)
+	}
+
+	// remove the address from the map so we can check if there are any
+	// missing addresses
+	m.v.deleteAccountHash(address)
+
+	return payloads, nil
+}
+
+func (m *CadenceDataValidationMigrations) hashAccountCadenceValues(
+	address common.Address,
+	payloads []*ledger.Payload,
+) ([]byte, error) {
+	hasher := newHasher()
+	mr, err := newMigratorRuntime(address, payloads)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, domain := range domains {
+		domainHash, err := m.hashDomainCadenceValues(mr, domain)
+		if err != nil {
+			return nil, fmt.Errorf("failed to hash storage domain %s : %w", domain, err)
+		}
+		_, err = hasher.Write(domainHash)
+		if err != nil {
+			return nil, fmt.Errorf("failed to write hash: %w", err)
+		}
+	}
+
+	return hasher.SumHash(), nil
+}
+
+func (m *CadenceDataValidationMigrations) hashDomainCadenceValues(
+	mr *migratorRuntime,
+	domain string,
+) ([]byte, error) {
+	hasher := newHasher()
+	var storageMap *interpreter.StorageMap
+	err := capturePanic(func() {
+		storageMap = mr.Storage.GetStorageMap(mr.Address, domain, false)
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get storage map: %w", err)
+	}
+	if storageMap == nil {
+		// no storage for this domain
+		return nil, nil
+	}
+
+	hashes := make(sortableHashes, 0, storageMap.Count())
+
+	iterator := storageMap.Iterator(util.NopMemoryGauge{})
+	for {
+		key, value := iterator.Next()
+		if key == nil {
+			break
+		}
+
+		h, err := m.recursiveString(value, hasher)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert value to string: %w", err)
+		}
+
+		hasher.Reset()
+
+		hashes = append(hashes, h)
+	}
+
+	return hashes.SortAndHash(hasher)
+}
+
+func (m *CadenceDataValidationMigrations) recursiveString(
+	value interpreter.Value,
+	hasher hash.Hasher,
+) ([]byte, error) {
+
+	var s string
+	err := capturePanic(
+		func() {
+			s = value.RecursiveString(interpreter.SeenReferences{})
```

Comment:
Not 100% if the string description of all values includes all data of the value.

Reply:
That file (cadence_data_validation.go) isn't used any more.  

So maybe the file can be removed since we're using cadence_value_validation.go instead.